cupyx/distributed/_comm.py

```diff
@@ -0,0 +1,206 @@
+import os
+
+import cupy
+from cupy.cuda import nccl
+from cupyx.distributed import _store
+
+_nccl_dtypes = {cupy.int8: nccl.NCCL_INT8,
+                cupy.uint8: nccl.NCCL_UINT8,
+                cupy.int32: nccl.NCCL_INT32,
+                cupy.uint32: nccl.NCCL_UINT32,
+                cupy.int64: nccl.NCCL_INT64,
+                cupy.uint64: nccl.NCCL_UINT64,
+                cupy.float16: nccl.NCCL_FLOAT16,
+                cupy.float32: nccl.NCCL_FLOAT32,
+                cupy.float64: nccl.NCCL_FLOAT64,
+                # Size of array will be doubled
+                cupy.complex64: nccl.NCCL_FLOAT32,
+                cupy.complex128: nccl.NCCL_FLOAT64}
+
+
+_nccl_ops = {'sum': nccl.NCCL_SUM,
+             'prod': nccl.NCCL_PROD,
+             'max': nccl.NCCL_MAX,
+             'min': nccl.NCCL_MIN}
+
+
+class NCCL:
+    def __init__(self, n_devices, rank):
```

Comment:
Usually `rank` is from MPI or multiprocessing. If we have our own TCPStore that can bypass any of these bootstrapping mechanism, I suppose we should also come up with a way to determine `rank`?

Reply:
Yeah, I've been banging my head with this and an easy way to determine the rank 0 process automatically since it is the one that starts the tcpstore. But my idea required using filesystem locks, matching the hostname and doing several other tricks ... In the end I just thought of manually assigning the ranks as pytorch does :D.

Reply:
I will do this in a follow up PR that boots and agent per each node in a similar way to pytorch.